### PR TITLE
IWTomics R/Bioconductor tool

### PR DESCRIPTION
**Pull request by fabio-cumbo**
IWTomics R/Bioconductor tool for Galaxy

### CODE REVIEW

File: tools/iwtomics/.shed.yml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+name: iwtomics
+owner: fabio
```
Comment (bgruening) [inline tools/iwtomics/.shed.yml]: `iuc` if you like.

File: tools/iwtomics/loadandplot.xml [hunk 2]
```diff
@@ -0,0 +1,357 @@
+<tool id="iwtomics_loadandplot" name="IWTomics Load" version="@VERSION@.0">
+  <description>Smooth and Plot</description>
+  <macros>
+    <import>macros.xml</import>
+  </macros>
+  <expand macro="requirements" />
+  <stdio>
+    <exit_code range="255" level="fatal" description="Missing IWTomics package. Please be sure to have it installed before using this tool." />
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: Is this mapping needed or are you creating the nice error description also at tool level? If so you can remove this I guess and use `detect_errors`

File: tools/iwtomics/loadandplot.xml [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    <exit_code range="34" level="fatal" description="Overlapping windows in feature file." />
+    <exit_code range="35" level="fatal" description="Windows in feature files do not cover all regions in region files." />
+    <exit_code range="36" level="fatal" description="All windows in a feature file must have the same size." />
+    <exit_code range="40" level="fatal" description="An error has occurred smoothing the data. Please try again." />
+    <regex match="Error" source="both" level="fatal" description="An error has occurred. Please try again." />
+  </stdio>
+
+  <command>
+<![CDATA[
+    Rscript $__tool_directory__/loadandplot.R outrdata="${outrdata}" outregions="${outregions}" outfeatures="${outfeatures}" outpdf="${outpdf}"
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: quote -> $__tool_directory__/loadandplot.R

Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: please single_quote everything not double quotes.

File: tools/iwtomics/loadandplot.xml [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    <!--
+    to print the stack add the following line at the end of the command and enable the corresponding entry in output
+    2> "${stackerr}"
+    -->
+  </command>
+
+  <inputs>
+    <!-- regions -->
+    <param format="tabular" name="regions" type="data" label="Select region files" multiple="true" help="" />
+    <repeat name="regionsheaderrepeat" title="Header file for regions (optional)" min="0" max="1" help="Tabular file. Column 1: names of the region files; Column 2: unique identifiers of the region datasets (spaces and special characters are not allowed); Column 3: name of the region datasets to be used in the output.">
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: min=0 max=1? Is the repeat really needed?

File: tools/iwtomics/loadandplot.xml [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+      <param format="tabular" name="regionsheader" type="data" label="Select header file" />
+    </repeat>
+
+    <!-- features -->
+    <param format="tabular" name="features" type="data" label="Select feature files" multiple="true" help="" />
+    <repeat name="featuresheaderrepeat" title="Header file for features (optional)" min="0" max="1" help="Tabular file. Column 1: names of the feature files; Column 2: unique identifiers of the features (spaces and special characters are not allowed); Column 3: name of the features to be used in the output.">
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: same here ...

File: tools/iwtomics/macros.xml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+<macros>
+    <token name="@VERSION@">1.0.0</token>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="3.3.1">R</requirement>
```
Comment (bgruening) [inline tools/iwtomics/macros.xml]: r-base

File: tools/iwtomics/loadandplot.xml [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+1. Simulated_data_
+2. ETn_data_
+
+.. _Simulated_data: https://usegalaxy.org/u/fabio-cumbo/h/iwtomics-example
+.. _ETn_data: https://usegalaxy.org/u/fabio-cumbo/h/iwtomics-etn-example
+  ]]></help>
+
+  <citations>
+    <citation type="bibtex">
+      @ARTICLE{
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: This could be added to the macro?

**Comment by fabio-cumbo** [comment]
Thanks @bgruening I will apply your suggestions asap!

File: packages/package_galaxy_json_collect_data_source/tool_dependencies.xml [hunk 8]
```diff
@@ -0,0 +1,26 @@
+<?xml version='1.0' encoding='utf-8'?>
+<tool_dependency>
+    <package name="python" version="2.7.10">
+        <repository name="package_python_2_7_10" owner="iuc" />
+    </package>
+    <package name="galaxy_json_collect_data_source" version="1.0.0">
+        <install version="1.0">
+            <actions>
+                <action type="download_by_url" target_filename="json_collect_data_source.py" sha256sum="c3035964913a8bfe4f196b133f9cde38bf220b97018737a4b761acc95ebf4e1e">https://raw.githubusercontent.com/fabio-cumbo/galaxy-json-collect-data-source/master/json_collect_data_source.py</action>
```
Comment (fabio-cumbo) [inline packages/package_galaxy_json_collect_data_source/tool_dependencies.xml]: Removed, it should not be here. I have create this pull request [#1363](https://github.com/galaxyproject/tools-iuc/pull/1363)

File: tools/iwtomics/loadandplot.xml (hunk 2 above)
Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: This is a set of three R script. I do not want to see the R warning and error messages, so I have redirected the standard error of the tools to ```>& /dev/null``` handling every single exception that could be thrown returning a specific exit code. This is why I need this mapping.

**Comment by fabio-cumbo** [comment]
All revisions applied, thanks @bgruening

File: tools/iwtomics/loadandplot.xml (hunk 2 above)
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: I think this can be done with https://docs.galaxyproject.org/en/master/dev/schema.html#id5

Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: Removed the ```stdio``` block in the tool XML schemas and created better error descriptions directly inside the R scripts

**Comment by fabio-cumbo** [comment]
@bgruening could you please check if I have correctly applied all your revisions?

File: tools/iwtomics/loadandplot.xml [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+  <inputs>
+    <!-- regions -->
+    <param format="tabular" name="regions" type="data" label="Select region files" multiple="true" help="" />
+    <section name="regionsheadersection" title="Insert header file for regions (optional)" expanded="false" help="Tabular file. Column 1: names of the region files; Column 2: unique identifiers of the region datasets (spaces and special characters are not allowed); Column 3: name of the region datasets to be used in the output.">
+      <param format="tabular" name="regionsheader" type="data" label="Select header file" optional="true" />
+    </section>
+
+    <!-- features -->
+    <param format="tabular" name="features" type="data" label="Select feature files" multiple="true" help="" />
+    <section name="featuresheadersection" title="Insert header file for features (optional)" expanded="false" help="Tabular file. Column 1: names of the feature files; Column 2: unique identifiers of the features (spaces and special characters are not allowed); Column 3: name of the features to be used in the output.">
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: Can you explain the reasoning why this and several other are in a section? Do you want to mark these input files as optional?

File: tools/iwtomics/loadandplot.xml [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+      <conditional name="conditionalplottype">
+        <!-- plot type -->
+        <param name="plottype" type="select" label="Plot type" help="">
+          <option value="boxplot">Pointwise quantile curves (boxplot)</option>
+          <option value="curves">Curves (aligned)</option>
+        </param>
+        <!-- conditional choice: plottype=boxplot -->
+        <when value="boxplot">
+          <section name="probabilitiessection" title="Probabilities" expanded="True" help="Probabilities corresponding to the quantile curves to be drawn.">
+            <param name="prob0" size="3" type="float" value="0.25" min="0.0" max="1.0" label="Probability" />
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: You can only use one repeat but set the min=3 or something like that.

File: tools/iwtomics/loadandplot.xml [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+          <when value="scale">
+            <param name="scalegrid" type="integer" value="0" min="0" label="Number of grid points" help="Number of equally-spaced grid points over which the smoothed curves are evaluated." />
+          </when>
+        </conditional>
+
+        <param name="splinesdegree" type="integer" value="3" min="0" label="Degree of B-splines" help="" />
+        <param name="splinesdistknots" type="integer" value="10" min="2" label="Distance between nodes (approximate)" help="" />
+
+        <!-- fill gaps -->
+        <param name="fillgaps" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="True" label="Use smoothing to fill gaps (NA measurements)" help="" />
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: Could be a marco?

File: tools/iwtomics/loadandplot.xml [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+        <param name="locpolydegree" type="integer" value="3" min="0" label="Degree of local polynomial" help="" />
+
+        <!-- fill gaps -->
+        <param name="fillgaps" type="boolean" truevalue="TRUE" falsevalue="FALSE" checked="True" label="Use smoothing to fill gaps (NA measurements)" help="" />
+      </when>
+      <!-- conditional choice: smoothing=kernel -->
+      <when value="kernel">
+        <!-- conditional region alignment -->
+        <conditional name="kernelconditionalscale">
+          <param name="alignment" type="select" label="Region alignment" help="">
```
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: Could this also be a macro? It is repeated below and above again, correct?

File: tools/iwtomics/macros.xml [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+    <token name="@VERSION@">1.0.0</token>
+    <xml name="requirements">
+        <requirements>
+            <requirement type="package" version="3.3.1">r-base</requirement>
+            <requirement type="package" version="@VERSION@">bioconductor-iwtomics</requirement>
+        </requirements>
+    </xml>
+    <xml name="citations">
+        <citations>
+            <citation type="bibtex">
```
Comment (bgruening) [inline tools/iwtomics/macros.xml]: if you like you can specify the doi `10.1080/10485252.2017.1306627` and remove the complicated bibtex

File: tools/iwtomics/plotwithscale.xml [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+      groupby="'${plotsum.conditionalgroupby.groupby}'"
+      #if $plotsum.conditionalgroupby.groupby == "test":
+        summaryalpha='${plotsum.conditionalgroupby.testalphaplot}'
+        only_significant='${plotsum.conditionalgroupby.testonlysig}'
+      #elif $plotsum.conditionalgroupby.groupby == "feature":
+        summaryalpha='${plotsum.conditionalgroupby.featurealphaplot}'
+        only_significant='${plotsum.conditionalgroupby.featureonlysig}'
+      #end if
+
+      >& /dev/null
```
Comment (bgruening) [inline tools/iwtomics/plotwithscale.xml]: Is this still needed?

File: tools/iwtomics/plotwithscale.xml [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+      <repeat name="thresholdontestscale" title="Threshold on test scale" min="1">
+        <param name="test" type="data_column" data_ref="testids" numerical="False" label="Test ID" multiple="True" use_header_names="True" />
+        <param name="feature" type="data_column" data_ref="featureids" numerical="False" label="Feature ID" multiple="True" use_header_names="True" />
+        <param name="scale" type="integer" value="0" min="0" label="Scale" />
+      </repeat>
+    </section>
+
+    <!-- plot IWTomics results -->
+    <section name="plotres" title="Plot IWTomics test results" expanded="True">
+      <!-- alpha -->
```
Comment (bgruening) [inline tools/iwtomics/plotwithscale.xml]: This also looks familiar to the other tool - marcofy?

File: tools/iwtomics/loadandplot.xml (hunk 10 above)
Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: The problem here is that I need to set different default values for each of these 'prob*' params. Using a repeat with min=3 just create three identical params, correct? Do you have any suggestion that allow me to use the repeat with at least three params and set a specific value for each of them?

**Comment by fabio-cumbo** [comment]
Thank you for your reviews @bgruening. This is my first attempt to create a Galaxy tool, so any suggestion is really appreciated!

File: tools/iwtomics/loadandplot.xml (hunk 9 above)
Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: Correct, these input files have to be optional. My intent here is to "hide" the field and let the user decide to activate it or not.

Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: I guess the currently more common way is to specify optional="true" inside the data param.

File: tools/iwtomics/loadandplot.xml (hunk 10 above)
Comment (bgruening) [inline tools/iwtomics/loadandplot.xml]: This is afaik not possible, so leave it like it is. Sorry for the noise.

File: tools/iwtomics/loadandplot.xml (hunk 9 above)
Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: Removed the section and set optional attribute to true

File: tools/iwtomics/loadandplot.xml (hunk 12 above)
Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: Done

File: tools/iwtomics/loadandplot.xml (hunk 11 above)
Comment (fabio-cumbo) [inline tools/iwtomics/loadandplot.xml]: Done

File: tools/iwtomics/macros.xml (hunk 13 above)
Comment (fabio-cumbo) [inline tools/iwtomics/macros.xml]: bibtex removed

File: tools/iwtomics/plotwithscale.xml (hunk 14 above)
Comment (fabio-cumbo) [inline tools/iwtomics/plotwithscale.xml]: right, I forgot to delete it

File: tools/iwtomics/plotwithscale.xml (hunk 15 above)
Comment (fabio-cumbo) [inline tools/iwtomics/plotwithscale.xml]: Done. Put the "plotsum" section in macro.xml too.

**Comment by fabio-cumbo** [comment]
@bgruening the code looks more readable now, thanks for your suggestions

**Comment by fabio-cumbo** [comment]
@bgruening no other comments in more than a month, I think that it is ready. I talked to @nekrut, this tool can be integrated in the main instance of Galaxy @davebx

**Comment by bgruening** [comment]
Thanks for the reminder - totally forget about it.
Thanks again for all your work here, in Bioconda and now also conda-forge.

**Pull request merged**